README.md
MatlabPlotter
=============
[![Build Status](https://travis-ci.org/pirminschmid/MatlabPlotter.svg?branch=master)](https://travis-ci.org/pirminschmid/MatlabPlotter)

A simple C++ class that writes Matlab .m programs as output of calculations e.g. using the Eigen C++ library. 

This may be useful if you do not want to add additional dependencies to your program or write the plotting code by yourself, but you know that all involved parties have Matlab installed on their computers.

Note: Due to the use of templates, any container type (e.g. vector, array,...) of any library can be used as input vector for plot() and print\_row\_vector(), as long as the container supports the standard methods size() and operator\[\]\(\), and the embedded scalar type can be printed by cout << and is suitable for plotting in Matlab. 

Current version v0.5 (2015-11-29). See [changelog][changelog] for details. [Feedback][feedback] welcome.


Usage
-----

Include matlab\_plotter.h into your C++ program. It needs the standard library of C++11. Use methods as shown in the examples [natcsi.cpp][example] ([folder][folder]) and [legendre.cpp][example2] ([folder][folder2]).

Note: the examples need [Eigen][eigen] to be installed on your system. You may use cmake for building.
However, the plotter can be used independently of it.


License
-------

Copyright (c) 2015 Pirmin Schmid, [MIT license][license].


[changelog]:https://github.com/pirminschmid/MatlabPlotter/tree/master/CHANGELOG.md
[example]:https://github.com/pirminschmid/MatlabPlotter/tree/master/example/natcsi.cpp
[folder]:https://github.com/pirminschmid/MatlabPlotter/tree/master/example
[example2]:https://github.com/pirminschmid/MatlabPlotter/tree/master/example2/legendre.cpp
[folder2]:https://github.com/pirminschmid/MatlabPlotter/tree/master/example2
[eigen]:http://eigen.tuxfamily.org
[license]:https://github.com/pirminschmid/MatlabPlotter/tree/master/LICENSE
[feedback]:mailto:mailbox@pirmin-schmid.ch?subject=MatlabPlotter


CHANGELOG.md

CHANGELOG
=========

v0.5 (2015-11-29)
-----------------
- summarizes v0.4.1 to 0.4.3
- minor changes in example2
- improved README
- added automatic testing by travis-ci
- some minor cosmetic changes

v0.4 (2015-11-23)
-----------------
- added title()
- removed restrictive type assertion at the beginning of plot() and print_row_vector().
  If these functions are used with wrong input type, the error messages will be more cryptic
  but this gives more flexibility for additional types that can use these templates
- new example2: Legendre polynomials and Gauss-Legendre quadrature  

v0.3 (2015-11-20)
-----------------
- plot() rewritten as templated method (no API change)
- print_row_vector() is now public
- added subplot()

v0.2 (2015-11-16)
-----------------
- added semilog and loglog plots
- added axis labels
- std::strings
- legend() changed to variadic template style (notice: API change)

v0.1 (2015-11-05)
-----------------
- Hello world!
- first draft -- warning: API may/will change
- figures
- plotting for C++ and Eigen vectors
- plot function in given interval
- classic style variadic function signature used for legend()
- thus: C style strings (char *)
- first example: Natural cubic spline interplolation

testing/run_examples.sh
#!/bin/bash
./natcsi >natcsi.m
./legendre >legendre.m


testing/build_examples.sh
#!/bin/bash
cd ../example/build
cmake ..
make
cp natcsi ../../testing/
cd ../../example2/build
cmake ..
make
cp legendre ../../testing/
cd ../../testing


example/natcsi.cpp
/*  example for MatlabPlotter: Natural cubic splines (core problem)

	An in-class exercise using equations for the cubic splines as
	shown in the lecture/manuscript for Numerical Methods for CSE
	by Prof. R. Hiptmair, ETH Zürich

	Include the Eigen3 library as shown in documentation for Eigen3.

	use piping to store the .m file. Example call:
	natcsi >example.m

	given nodes (t_i,y_i) for i = 0,...,n

	and using eq 3.5.5, the calculation of y = s(x) boils down to
	- finding the correct interval [t_j-1,t_j] for x
	- apply s|[t_j-1,t_j](x) = s_j(x) =
		y_j-1 * (1 - 3 * tau^2 + 2 * tau^3) +
		y_j   * (3 * tau^2 - 2 * tau^3) +
		m_j-1 * (tau - 2 * tau^2 + tau^3) +
		m_j   * (-tau^2 + tau^3)

		with
		m1_j = m_j-1 = h_j * c_j-1,
		m2_j = m_j   = h_j * c_j
		
		c_j   = s'(t_j)
		
		h_j   = t_j - t_j-1

		tau   = (t - t_j-1) / h_j

	current runtime:
	- initialization: O(n), Thomas algorithm for solving the tridiagonal system matrix
	- evaluation:
		* worst case:        O(log n)
		* best/typical case: O(1)
	
	space requirements: O(n) all the time             

	output: is given again as a Matlab script that
	allows plotting all data in Matlab.

	v1.1.1 2015-11-05 / 2015-11-29 Pirmin Schmid 
*/

#include <cassert>
#define _USE_MATH_DEFINES
#include <cmath>
#include <vector>
#include <Eigen/Dense>
#include "matlab_plotter.h"

using namespace std;
using namespace Eigen;

// cubic spline interpolant with natural boundaries
class NatCSI {
public:
	
	// build cubic spline interpolant with natural boundaries
	// input:  t  nodes of the grid for pairs (t_i, y_i), sorted!
	//         y  values y_i at t_i for pairs (t_i, y_i), sorted!
	// input in std::vector<double> type
	NatCSI(const vector<double>& t, const vector<double>& y) {
		// map std::vector<double> to a VectorXd
		VectorXd tmp_t = VectorXd::Map(t.data(), t.size());
		VectorXd tmp_y = VectorXd::Map(y.data(), y.size());
		
		// store a copy of the data
		this->t = tmp_t;
		this->y = tmp_y;
		
		compute_coefficients();
	}
	
	// build cubic spline interpolant with natural boundaries
	// input:  t  nodes of the grid for pairs (t_i, y_i), sorted!
	//         y  values y_i at t_i for pairs (t_i, y_i), sorted!
	// input in Eigen::VectorXd type
	NatCSI(const VectorXd& t, const VectorXd& y) {
		// store a copy of the input vectors
		this->t = t;
		this->y = y;
		
		compute_coefficients();
	}
	
	// interpolant evaluation at x
	// input:   x value where to evaluate the spline
	// return:  y = s(x) if x in defined range of s(x)
	//          or 0.0 otherwise
	double operator() (double x) {
		// out of defined interval
		if(x < min || max < x) {
			return 0.0;
		}
		
		// get proper j, using a heuristic
		// 1) check whether x is in last range, O(1)
		// 2) check whether x is in next range, O(1)
		// 3) use binary search tree, O(log n)
		if(!in_range(x, last_j)) {
			if(last_j < n && in_range(x, last_j+1)) {
				// note: the second test is only applied if last_j actually allows it
				last_j++;
			}
			else {
				find_index(x);
			}
		}
		
		// apply function in interval j
		int j = last_j; // just for convenience
		
		double tau  = (x - t[j-1]) / h[j];
		double tau2 = tau * tau;
		double tau3 = tau2 * tau;
		
		double result = y[j-1] * (1.0 - 3.0 * tau2 + 2.0 * tau3);
		result += y[j] * (3.0 * tau2 - 2.0 * tau3);
		result += m1[j] * (tau - 2.0 * tau2 + tau3);
		result += m2[j] * (-tau2 + tau3);
		return result;
	}
	
private:
	// we need to store the vectors t and y
	// and a vector for m and h as described above
	// m is calculated after having solved a LSE
	// A * c = d as described below (and in the manuscript)
	
	VectorXd t, y, h, m1, m2;
	
	// number of segments of the spline function
	// note: n is a valid index for the vectors. Their size is (n+1).
	int n = 0;
	
	// valid range for x
	double min = 0.0;
	double max = 0.0;
	
	// since most requests for x calculations are in vicinity
	// of the prior one (or even sorted), remembering
	// the last requested section will help to bring
	// evaluation time to O(1) compared to O(log n) if only
	// a binary search tree is used.
	// the binary search tree is currently used as a fallback
	// in case the current partition/segment or the next one
	// has not been successful.
	int last_j = 1;
	
	// returns true if x is in interval j [t_j-1, t_j]
	bool in_range(const double x, const int j) const {
		assert(1 <= j && j <= n);
		return t[j-1] <= x && x <= t[j];
	}
	
	// sets last_j to  index of matching interval j [t_j-1, t_j] if x in range
	// note: x had to be tested before whether it is actually within the range of the spline function
	// note: optimizations (first test whether last interval was already ok or test the next
	// interval have already been done when one calls this function)
	// finds next j in O(log n) time
	void find_index(const double x) {
		assert(min <= x && x <= max);
		
		int j = last_j;
		int a = 1;
		int b = n;
		bool found = false;
		while(!found) {
			if(x < t[j-1]) {
				b = j;
				j = (a+b)/2;
			}
			else if(t[j] < x) {              
				if(a == b-1) {
					// special case needed due to integer division rounding down
					j = b;
					found = true;
				}
				else {
					a = j;
					j = (a+b)/2;
				}
			}
			else {
				// match
				found = true;
			}
		}
		
		last_j = j;
	}
	
	void compute_coefficients() {
		if(t.size() != y.size()) {
			// only minimalistic error management
			cout << "Error: size of t and y must be identical." << endl;
			exit(1);
		}
		
		// n
		n = (int)t.size() - 1;
		
		// min/max
		min = t[0];
		max = t[n];
		
		// h_j = t_j - t_j-i
		h.resize(n + 1);
		h.segment(1, n) = t.tail(n) - t.head(n);
		
		// build a virtual (n+1)x(n+1) system matrix A; however, only the diagonal vectors will actually be built
		// A = diag(a) + diag(b,1) + diag(b,-1), with size(a) = n+1 and size(b) = n
		
		// b_i = 1/h_i+1 for i=0 to n-1
		VectorXd b = h.tail(n).cwiseInverse();
		
		// a_i = 2/h_i + 2/h_i+1 for i=1 to n-1
		// a_0 = 2/h_1 (see natural cubic spline)
		// a_n = 2/h_n (see natural cubic spline)
		VectorXd a(n + 1);
		a.segment(1, n-1) = 2.0 * (h.segment(1, n-1).cwiseInverse() + h.tail(n-1).cwiseInverse());
		a[0] = 2.0 / h[1];
		a[n] = 2.0 / h[n];
		
		// right hand side d for A*c = d
		// for i=1 to n-1
		// d_i = 3.0 * ( (y_i - y_i-1) / h2_i  +  (y_i+1 - y_i) / h2_i+1 )
		// d_0 = 3.0 * (y_1 - y_0) / h2_1
		// d_n = 3.0 * (y_n - y_n-1) / h2_n
		
		VectorXd h2 = h.cwiseProduct(h); // h^2
		
		// note: y_i - y_i-1 is stored in delta_y[i-1]
		VectorXd delta_y = y.tail(n) - y.head(n);
		
		// quotient is either left or right side inside the brackets
		VectorXd quotient = delta_y.cwiseQuotient(h2.tail(n));
		
		// build d
		VectorXd d(n + 1);
		d.segment(1, n-1) = quotient.head(n-1) + quotient.tail(n-1);
		d[0] = delta_y[0] / h2[1];
		d[n] = delta_y[n-1] / h2[n];
		d = 3.0 * d;
		
		// Solution of this LSE with n+1 equations and n+1 unknowns c0 to cn
		// using the Thomas algorithm for tridiagonal matrices, needs O(n)
		// typical description with vectors b1..bn, a2..an, c1..cn-1
		// 'b' is a here
		// 'a' and 'c' are b here
		// 'x' is our c
		// note: different indices, too; also we have size n+1 instead of n
		
		// forward preparation
		VectorXd bprime(n);
		VectorXd dprime(n+1);
		
		// cprime -> bprime
		bprime[0] = b[0] / a[0];
		for(int i = 1; i < (n-1); ++i) {
			bprime[i] = b[i] / (a[i] - bprime[i-1] * b[i]);
		}
		
		// dprime
		dprime[0] = d[0] / a[0];
		for(int i = 1; i < n; ++i) {
			dprime[i] = (d[i] - dprime[i-1] * b[i]) / (a[i] - bprime[i-1] * b[i]);
		}
		
		// back substitution to get x, or c in our case
		VectorXd c(n + 1);
		c[n] = dprime[n];
		for(int i = n-1; i >= 0; --i) {
			c[i] = dprime[i] - bprime[i] * c[i+1];
		}
		
		// we do not actually need c but
		// m1_j = m_j-1 = h_j * c_j-1,
		// m2_j = m_j   = h_j * c_j
		m1.resize(n+1);
		m2.resize(n+1);
		m1.segment(1, n) = h.tail(n).cwiseProduct( c.head(n) );
		m2.segment(1, n) = h.tail(n).cwiseProduct( c.tail(n) );
	}
	
};

// just a quick implementation of the function
double f(const double x) {
	return exp(sin(2.0 * M_PI * x));
}

#define MIN -3
#define MAX 3
#define N_T 61
#define N_X 600

int main() {
	// initialization
	VectorXd t = VectorXd::LinSpaced(N_T, MIN, MAX);
	VectorXd y(N_T);
	for(int i = 0; i < N_T; ++i) {
		y[i] = f(t[i]);
	}
	
	NatCSI ncsi(t, y);
	
	// test 1
	VectorXd xx = VectorXd::LinSpaced(N_X, MIN, MAX);
	VectorXd yy(N_X);
	for(int i = 0; i < N_X; ++i) {
		yy[i] = ncsi(xx[i]);
	}
	
	// test 2. test the binary search in find_index()
	VectorXd y2(N_T);
	for(int i = N_T-1; i >= 0; --i) {
		y2[i] = ncsi(t[i]);
	}
	
	// test 3. test border case (last node)
	// make sure that the 1st AND the last node are highlighted
	// in the plot (last node: see special case handling due to
	// integer division in find_index() )
	VectorXd x3(2);
	VectorXd y3(2);
	x3 << t[0], t[N_T-1];
	for(int i = 0; i < x3.size(); ++i) {
		y3[i] = ncsi(x3[i]);
	}
	
	// test 4. test std::vector<double> (no legend item)
	vector<double> x4 = {MIN, MAX};
	vector<double> y4 = {1.0, 1.0};

	// plot results
	MatlabPlotter p;
	p.comment("Example: natural cubic splines");
	p.comment("Code generated by natcsi.cpp");
	p.figure("Natural CSI");
	p.plot_fx("exp(sin(2.*pi.*x))", MIN, MAX, N_X, "b-");
	p.hold();
	p.plot(t, y, "b*");
	p.plot(xx, yy, "r--");
	p.plot(t, y2, "ro");
	p.plot(x3, y3, "g*");
	p.plot(x4, y4, "k:");
	p.xylabels("x", "y");
	p.legend("exp(sin(2*pi*x))", "t0 to tn", "Natural CSI", "check findindex()", "check corner cases");
	p.hold(false);

	// for testing / illustration purpose: semilog scale for y axis
	p.figure("Natural CSI. semilog scale", MatlabPlotter::SEMILOGY);
	p.plot_fx("exp(sin(2.*pi.*x))", MIN, MAX, N_X, "b-");
	p.hold();
	p.plot(t, y, "b*");
	p.plot(xx, yy, "r--");
	p.plot(t, y2, "ro");
	p.plot(x3, y3, "g*");
	p.plot(x4, y4, "k:");
	p.xylabels("x", "y (log scale)");
	p.legend("exp(sin(2*pi*x))", "t0 to tn", "Natural CSI", "check findindex()", "check corner cases");
	p.hold(false);
}


matlab_plotter.h
/*  matlab_plotter.h

	A simple interface to plot C++ results by Matlab. Basically, a .m file is composed.
	This is an include-only library without dependency except the C++ standard library
	(C++11). If you are using the Eigen3 library, specific plot methods are offered for
	Eigen vectors, too.

	The 'program' is printed to stdout. Use piping to store it in a .m file.
	Do not write other output to stdout when you intend to use this method.

	Yes, it's trivial, but it fit the bill when needed.

	Feedback welcome: mailbox@pirmin-schmid.ch

	v0.5 2015-11-05 / 2015-11-29; early development, interface may change

	get latest version from: https://github.com/pirminschmid/MatlabPlotter

	(c) 2015 Pirmin Schmid, MIT License

	license text: https://github.com/pirminschmid/MatlabPlotter/tree/master/LICENSE
*/


#ifndef _MATLAB_PLOTTER_H
#define _MATLAB_PLOTTER_H

#include <iostream>
#include <string>

class MatlabPlotter {
public:
	MatlabPlotter() {
		// currently nothing
	}

	~MatlabPlotter() {
		// currently nothing
	}


	//--- public types

	enum PlotType {
		LINEAR,
		SEMILOGY,
		SEMILOGX,
		LOGLOG
	};


	//--- public methods

	// writes a comment
	void comment(const std::string& comment) const {
		std::cout << "% " << comment << std::endl;
	}


	// starts a new figure with name title and currently set plotType
	void figure(const std::string& title) const {
		std::cout << "figure('Name','" << title << "');" << std::endl;
	}


	// starts a new figure with name title and sets plotType for all future plots
	void figure(const std::string& title, const PlotType newType) {
		plotType = newType;
		figure(title);
	}


	// writes the Matlab commands hold on or hold off
	void hold(const bool on = true) const {
		if(on) {
			std::cout << "hold on;" << std::endl;
		}
		else {
			std::cout << "hold off;" << std::endl;
		}
	}


	// prints the legend. This is a variadic function using C++ variadic template functionality
	// input: one or more variables, typically strings (type std::string)
	// NOTICE: This method does not have the first parameter 'int count' anymore (see v0.1).
	//         Please adjust your code if needed. Sorry for the inconvenience.
	template <typename T>
	void legend(T s) {
		if(lps == INACTIVE) {
			// only one argument
			std::cout << "legend('" << s << "');" << std::endl;
		}
		else {
			// last of a series of arguments
			std::cout << "'" << s << "');" << std::endl;
			lps = INACTIVE;
		}
	}

	template<typename T, typename... Args>
	void legend(T first, Args... more) {
		if(lps == INACTIVE) {
			// prefix
			std::cout << "legend(";
			lps = ACTIVE;
		}

		// first / or next argument
		std::cout << "'" << first << "',";

		// handle more arguments
		legend(more...);
	}


	// plots data given in vectors x and y
	// input: x and y    vectors of scalars, typically stored in std::vector<double>, Eigen::VectorXd, or Eigen::ArrayXd
	//                   the Vector type has to offer size() and access operator[]
	//                   the embedded scalar type must be printable by cout <<, and must be a suitable number type for
	//                   plotting in Matlab
	//        style      a string with a Matlab style description
	template<typename Vector>
	void plot(const Vector& x, const Vector& y, const std::string& style) const {
		if(x.size() != y.size()) {
			std::cout << "ERROR - plot(): vectors x and y must have equal size." << std::endl;
			return;
		}

		print_row_vector("x", x);
		print_row_vector("y", y);
		std::cout << plotCommand() << "(x,y,'" << style << "');" << std::endl;
	}


	// prints a vector x as Matlab row vector
	// input: name       string with a suitable variable name in Matlab
	// input: values     vector of scalars, typically stored in std::vector<double>, Eigen::VectorXd, or Eigen::ArrayXd
	//                   the Vector type has to offer size() and access operator[]
	//                   the embedded scalar type must be printable by cout <<, and must be a suitable number type for
	//                   plotting in Matlab
	template<typename Vector>
	void print_row_vector(const std::string& name, const Vector& values) const {
		bool first = true;
		std::cout << name << " = [";
		int n = values.size();
		for(int i = 0; i < n; ++i) {
			if(first) {
				first = false;
				std::cout << values[i];
			}
			else {
				std::cout << ", " << values[i];
			}
		}
		std::cout << "];" << std::endl;
	}


	// plots the function f using linspace for x with min, max and n=number of points
	// note: f must be described in Matlab syntax that allows vector-wise operations
	//       therefore, use .* ./ .^ instead of * / and ^, respectively
	//       you must use x as the variable name
	void plot_fx(const std::string& f, const double min, const double max, const int n, const std::string& style) const {
		std::cout << "f = @(x) " << f << ";" << std::endl;
		std::cout << "x = linspace(" << min << "," << max << "," << n << ");" << std::endl;
		std::cout << plotCommand() << "(x,f(x),'" << style << "');" << std::endl;
	}


	// prints any string text as is
	void raw(const std::string& text) const {
		std::cout << text << std::endl;
	}


	// subplot(m,n,p) command as defined by Matlab
	void subplot(const int m, const int n, const int p) const {
		std::cout << "subplot(" << m << "," << n << "," << p << ");" << std::endl;
	}

	// adds title
	// note: as in Matlab, title needs to be called after the first plot to be effective
	// (it adds a title to existing axes). This function here behaves identically on purpose.
	void title(const std::string& title) const {
		std::cout << "title('" << title << "');" << std::endl;
	}

	// adds x and y axis labels
	void xylabels(const std::string& xlabel, const std::string& ylabel) const {
		std::cout << "xlabel('" << xlabel << "');" << std::endl;
		std::cout << "ylabel('" << ylabel << "');" << std::endl;
	}

private:

	// use a tiny finite state machine to help controlling the variadic template methods for legend()
	enum LegendPrintState {
		INACTIVE,
		ACTIVE
	};

	LegendPrintState lps = INACTIVE;


	// control the plotting type/scale for the figure
	PlotType plotType = LINEAR;

	std::string plotCommand() const {
		switch(plotType) {
			case LINEAR:
				return "plot";
			case SEMILOGY:
				return "semilogy";
			case SEMILOGX:
				return "semilogx";
			case LOGLOG:
				return "loglog";
			default:
				return "error_wrong_plot_type";
		}
	}

};

#endif // _MATLAB_PLOTTER_H
